addon/-private/coordinator.js

```diff
@@ -0,0 +1,110 @@
+import Cache from './cache';
+import RecordQuery from './record-query';
+import RecordArrayQuery from './record-array-query';
+import { get } from '@ember/object';
+
+/*
+  I know how to retrieve queries from the cache, and also assemble queries that
+  are not in the cache but can be derived from them.
+*/
+export default class Coordinator {
+
+  constructor(store) {
+    this.store = store;
+    this.cache = new Cache();
+
+    // A materialized view of loaded includes from the cache's queries.
+    this.loadedIncludes = {};
+  }
+
+  recordQueryFor(type, id, params) {
+    let query = this.cache.getRecordQuery(type, id, params);
+
+    if (!query) {
+      query = this._assembleRecordQuery(type, id, params);
+      this._rememberRecordQuery(query);
+    }
+
+    return query;
+  }
+
+  recordArrayQueryFor(type, params) {
+    let query = this.cache.getRecordArrayQuery(type, params);
+
+    if (!query) {
+      query = this._assembleRecordArrayQuery(type, params);
+      this._rememberRecordArrayQuery(query);
+    }
+
+    return query;
+  }
```

Comment:
This feels like two classes? I'm seeing recordArray and recordQuery functions.

Reply:
I thought about having all `query*` methods and classes being one, and using multiple arity functions to switch between records and collections (similar to how `store.find()` used to work), but figured it'd be better to split out into explicit `record` and `recordArray` methods and classes for clarity sake. 

So, there's two types of queries: a `RecordQuery` (for a single model) and a `RecordArrayQuery` (for a collection). Coordinator knows how to assemble them, put them in the cache, etc. That's why there seems like there's two of everything.